metadata/src/main/java/org/apache/kafka/metadata/BrokerRegistration.java

```diff
@@ -152,13 +156,22 @@ public boolean fenced() {
         return fenced;
     }
 
-    public ApiMessageAndVersion toRecord() {
+    public boolean inControlledShutdown() {
+        return inControlledShutdown;
+    }
+
+    public ApiMessageAndVersion toRecord(MetadataVersion metadataVersion) {
         RegisterBrokerRecord registrationRecord = new RegisterBrokerRecord().
             setBrokerId(id).
             setRack(rack.orElse(null)).
             setBrokerEpoch(epoch).
             setIncarnationId(incarnationId).
             setFenced(fenced);
+
+        if (metadataVersion.isInControlledShutdownStateSupported()) {
+            registrationRecord.setInControlledShutdown(inControlledShutdown);
+        }
```

Comment:
Isn't is an error if `inControlledShutdown` is true and `metadataVersion` doesn't support this field? In other words I think we can just set this field is `inControlledShutdown` is `true`.

I suspect that we would get an `UnsupportedVersionException` when writing this record to the log or snapshot.

Reply:
Hmm. I am not sure about this. In the case of a "unclean" downgrade, we need to write the state with an older version and we could have `inControlledShutdown` set in the `BrokerRegistration` if we briefly used the metadata version supporting that. It seems better to me to be defensive here.

Reply:
Yeah. This is a question I have had it the back of my head for a while with respect to downgrades. [KIP-778](https://wiki.apache.org/confluence/pages/viewpage.action?pageId=188746840#KIP778:KRaftUpgrades-Downgrades) talks about generating and loading a new snapshot when the metadata.version is downgraded. The active controller is not allowed to load a snapshot so that means that the active controller will have to resign after generating a snapshot but before loading the new snapshot. cc @mumrah 

In other words, I think you are correct. We need to be able to generate a snapshot to a previous metadata version.